File: proxy-android/proxy/src/main/java/com/zeticai/melangelm/pipeline/ProxyPipeline.kt
package com.zeticai.melangelm.pipeline

import android.util.Log
import com.zeticai.melangelm.model.BlockReason
import com.zeticai.melangelm.model.ChatRequest
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import com.zeticai.melangelm.model.PipelineOnlyResult
import com.zeticai.melangelm.model.ProxyResult
import com.zeticai.melangelm.model.ProxySavings
import com.zeticai.melangelm.model.StageResult
import com.zeticai.melangelm.model.StageStatus
import com.zeticai.melangelm.stages.SummarizerStage
import com.zeticai.melangelm.upstream.UpstreamClient
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

private const val TAG = "ProxyPipeline"

/**
 * Orchestrates the full request lifecycle:
 *   1. Initialize stages (once)
 *   2. For each request: run request stages in order
 *   3. If not blocked: forward to upstream LLM
 *   4. Run response stages in reverse order (unwinding)
 *   5. Return result to caller
 */
class ProxyPipeline(
    private val stages: List<PipelineStage>,
    private val upstream: UpstreamClient
) {
    @Volatile private var initialized = false

    fun updateCompressionRatio(ratio: Double) {
        stages.filterIsInstance<SummarizerStage>().forEach { it.setCompressionRatio(ratio) }
    }

    suspend fun initialize(
        onStageReady: ((String) -> Unit)? = null,
        onStageProgress: ((String, Float) -> Unit)? = null
    ) {
        if (initialized) return
        coroutineScope {
            stages.map { stage ->
                async(Dispatchers.IO) {
                    Log.d(TAG, "Initializing stage: ${stage.name}")
                    val progressForStage: ((Float) -> Unit)? = onStageProgress?.let { callback ->
                        { progress: Float -> callback(stage.name, progress) }
                    }
                    runCatching { stage.initialize(progressForStage) }
                        .onSuccess { onStageReady?.invoke(stage.name) }
                        .onFailure { Log.e(TAG, "Stage ${stage.name} init failed", it) }
                }
            }.awaitAll()
        }
        initialized = true
        Log.i(TAG, "Pipeline initialized with ${stages.size} stage(s): ${stages.map { it.name }}")
    }

    suspend fun process(chatRequest: ChatRequest): ProxyResult = withContext(Dispatchers.IO) {
        if (!initialized) initialize()

        val request = ProxyRequest(chatRequest)
        val originalCharCount = chatRequest.messages.sumOf { it.content.length }

        // --- Request phase ---
        for (stage in stages) {
            runCatching { stage.processRequest(request) }
                .onFailure { e ->
                    Log.e(TAG, "Stage ${stage.name} threw during processRequest", e)
                    return@withContext ProxyResult.Error("Stage '${stage.name}' failed: ${e.message}", e)
                }
            if (request.isBlocked) {
                Log.i(TAG, "Request blocked by stage '${stage.name}': ${request.blockReason}")
                return@withContext ProxyResult.Blocked(BlockReason.MALICIOUS_PROMPT, stage.name)
            }
        }

        // --- Upstream ---
        val upstreamResponse = runCatching { upstream.send(request.toUpstreamRequest()) }
            .getOrElse { e ->
                Log.e(TAG, "Upstream call failed", e)
                return@withContext ProxyResult.Error("Upstream failed: ${e.message}", e)
            }

        // --- Response phase (reverse order so stages can "unwind") ---
        var response = upstreamResponse
        for (stage in stages.reversed()) {
            response = runCatching { stage.processResponse(request, response) }
                .getOrElse { e ->
                    Log.w(TAG, "Stage ${stage.name} threw during processResponse, skipping", e)
                    response
                }
        }

        val processedCharCount = request.messages.sumOf { it.content.length }
        val savings = ProxySavings(
            originalCharCount = originalCharCount,
            processedCharCount = processedCharCount,
            upstreamPromptTokens = response.usage?.promptTokens,
            upstreamCompletionTokens = response.usage?.completionTokens,
            costPerInputToken = ProxySavings.GPT_4O_MINI_INPUT_PER_TOKEN,
            costPerOutputToken = ProxySavings.GPT_4O_MINI_OUTPUT_PER_TOKEN
        )

        ProxyResult.Success(response, savings)
    }

    suspend fun processOnly(chatRequest: ChatRequest): PipelineOnlyResult = withContext(Dispatchers.IO) {
        if (!initialized) initialize()

        val request = ProxyRequest(chatRequest)
        val stageResults = mutableListOf<StageResult>()

        for (stage in stages) {
            val before = request.messages.toList()
            runCatching { stage.processRequest(request) }
                .onFailure { e ->
                    stageResults.add(StageResult(stage.name, StageStatus.ERROR, detail = e.message))
                    return@withContext PipelineOnlyResult(
                        isBlocked = false,
                        blockedBy = null,
                        blockReason = e.message,
                        processedMessages = request.messages,
                        stageResults = stageResults
                    )
                }
            val detail = request.metadata["${stage.name}.detail"] as? String
            if (request.isBlocked) {
                stageResults.add(StageResult(stage.name, StageStatus.BLOCKED, detail = detail))
                return@withContext PipelineOnlyResult(
                    isBlocked = true,
                    blockedBy = stage.name,
                    blockReason = request.blockReason,
                    processedMessages = request.messages,
                    stageResults = stageResults
                )
            }
            val changed = before != request.messages
            stageResults.add(StageResult(stage.name, if (changed) StageStatus.MODIFIED else StageStatus.PASSED, detail = detail))
        }

        PipelineOnlyResult(
            isBlocked = false,
            blockedBy = null,
            blockReason = null,
            processedMessages = request.messages,
            stageResults = stageResults
        )
    }
}

File: proxy-ios/Sources/MelangeLmProxy/Pipeline/ProxyPipeline.swift
//
//  ProxyPipeline.swift
//  MelangeLmProxy
//
//  Orchestrates the full request lifecycle through all stages.
//

import Foundation

/// Orchestrates the full request lifecycle:
///  1. Initialize all stages (once, lazily)
///  2. Run each stage's `processRequest` in order
///  3. If not blocked, forward to the upstream LLM
///  4. Run `processResponse` in reverse order (unwinding)
///  5. Return `ProxyResult` to the caller
actor ProxyPipeline {
    private let stages: [any PipelineStage]
    private let upstream: any UpstreamClient
    private var initialized = false

    init(stages: [any PipelineStage], upstream: any UpstreamClient) {
        self.stages = stages
        self.upstream = upstream
    }

    nonisolated func updateCompressionRatio(_ ratio: Double) {
        for stage in stages {
            if let summarizer = stage as? SummarizerStage {
                summarizer.setCompressionRatio(ratio)
            }
        }
    }

    func initialize(
        onStageReady: ((String) -> Void)? = nil,
        onStageProgress: ((String, Float) -> Void)? = nil
    ) async {
        guard !initialized else { return }
        await withTaskGroup(of: Void.self) { group in
            for stage in stages {
                group.addTask {
                    do {
                        let progressForStage: ((Float) -> Void)? = onStageProgress.map { callback in
                            return { progress in callback(stage.name, progress) }
                        }
                        try await stage.initialize(onProgress: progressForStage)
                        onStageReady?(stage.name)
                    } catch {
                        // logged below
                    }
                }
            }
        }
        initialized = true
        log("Pipeline ready with \(stages.count) stage(s): \(stages.map(\.name).joined(separator: ", "))")
    }

    func process(_ chatRequest: ChatRequest) async -> ProxyResult {
        if !initialized { await initialize() }

        let request = ProxyRequest(originalRequest: chatRequest)

        // --- Request phase ---
        for stage in stages {
            do {
                try await stage.processRequest(request)
            } catch {
                log("Stage '\(stage.name)' threw during processRequest: \(error)", level: .error)
                return .failure(message: "Stage '\(stage.name)' failed: \(error.localizedDescription)", error: error)
            }
            if request.isBlocked {
                log("Request blocked by '\(stage.name)': \(request.blockReason ?? "")")
                return .blocked(reason: .maliciousPrompt, stage: stage.name)
            }
        }

        // --- Upstream ---
        let upstreamResponse: ChatResponse
        do {
            upstreamResponse = try await upstream.send(request.toUpstreamRequest())
        } catch {
            log("Upstream failed: \(error)", level: .error)
            return .failure(message: "Upstream failed: \(error.localizedDescription)", error: error)
        }

        // --- Response phase (reverse order) ---
        var response = upstreamResponse
        for stage in stages.reversed() {
            do {
                response = try await stage.processResponse(request, response: response)
            } catch {
                log("Stage '\(stage.name)' threw during processResponse — skipping: \(error)", level: .warning)
            }
        }

        return .success(response)
    }

    /// Run only the on-device pipeline stages (no upstream call).
    /// Returns the processed request so the caller can inspect what each stage did.
    func processOnly(_ chatRequest: ChatRequest) async -> PipelineOnlyResult {
        if !initialized { await initialize() }

        let request = ProxyRequest(originalRequest: chatRequest)

        var stageResults: [StageResult] = []

        for stage in stages {
            let before = request.messages
            do {
                try await stage.processRequest(request)
            } catch {
                stageResults.append(StageResult(name: stage.name, status: .error(error.localizedDescription)))
                continue
            }
            let detailKey = "\(stage.name).detail"
            let detail = request.metadata[detailKey] as? String

            if request.isBlocked {
                stageResults.append(StageResult(name: stage.name, status: .blocked(request.blockReason ?? "blocked"), detail: detail))
                return PipelineOnlyResult(
                    isBlocked: true,
                    blockedBy: stage.name,
                    blockReason: request.blockReason,
                    processedMessages: request.messages,
                    stageResults: stageResults
                )
            }
            let changed = before != request.messages
            stageResults.append(StageResult(name: stage.name, status: changed ? .modified : .passed, detail: detail))
        }

        return PipelineOnlyResult(
            isBlocked: false,
            blockedBy: nil,
            blockReason: nil,
            processedMessages: request.messages,
            stageResults: stageResults
        )
    }

    // MARK: - Logging

    private enum LogLevel { case info, warning, error }

    private func log(_ message: String, level: LogLevel = .info) {
        #if DEBUG
        let prefix: String
        switch level {
        case .info:    prefix = "[MelangeLmProxy]"
        case .warning: prefix = "[MelangeLmProxy ⚠️]"
        case .error:   prefix = "[MelangeLmProxy ❌]"
        }
        print("\(prefix) \(message)")
        #endif
    }
}
